torch/testing/_internal/common_methods_invocations.py

```diff
@@ -5219,18 +5219,38 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
     return op(input.triu() if upper else input.tril(), upper)
 
 
-def reference_reduction_numpy(f):
+def reference_reduction_numpy(f, supports_keepdims=True):
     @wraps(f)
-    def wrapper(t: torch.Tensor, *args, dim=None, keepdim=False, **kwargs):
-        # NumPy reductions don't accept dim=0 for scalar inputs
-        if t.ndim == 0 and (dim == 0 or dim == -1):
-            dim = None
+    def wrapper(t: torch.Tensor, *args, **kwargs):
+        keys = kwargs.keys()
+
+        if 'dim' in keys:
+            dim = kwargs.pop('dim')
+
+            # NumPy reductions don't accept dim=0 for scalar inputs
+            kwargs['axis'] = None if t.ndim == 0 and (dim == 0 or dim == -1) else dim
+
+        if 'keepdim' in keys:
+            if not supports_keepdims:
+                # NumPy does not support keepdims kwarg so we manually unsqueeze result
+                if kwargs.pop('keepdim'):
+                    res = f(t.cpu().numpy(), *args, **kwargs)
+                    dims = kwargs.get('axis', list(range(t.ndim)))
+                    return np.expand_dims(res, dims)
+            else:
+                kwargs['keepdims'] = kwargs.pop('keepdim')
 
-        return f(t.cpu().numpy(), *args, axis=dim, keepdims=keepdim, **kwargs)
+        return f(t.cpu().numpy(), *args, **kwargs)
 
     return wrapper
 
 
+def reference_logsumexp(t, dim=None, keepdim=False):
```

Comment:
Cool reference; the version without SciPy may have numerical stability issues but I don't think our CI runs without SciPy. Another option would be to only define the reference if SciPy is available.